etcdserver/snapshot_store.go

```diff
@@ -93,29 +105,93 @@ func newSnapshotStore(dir string, kv dstorage.KV) *snapshotStore {
+// Otherwise, it returns the snapshot stored.
+//
+// The created snapshot is cleared from the snapshot store if it is
+// either unused after some time interval, or reported that it is used up.
+// The snapshot could be closed to become unreadable to be reported fast.
+//
+// snapshot store stores at most one snapshot at a time.
 // If raft state machine wants to send two snapshot messages to two followers,
 // the second snapshot message will keep getting snapshot and succeed only after
 // the first message is sent. This increases the time used to send messages,
 // but it is acceptable because this should happen seldomly.
 func (ss *snapshotStore) getSnap() (*snapshot, error) {
-	// If snapshotStore has some snapshot that has not been closed, it cannot
-	// request new snapshot. So it returns ErrSnapshotTemporarilyUnavailable.
-	if ss.snap != nil && !ss.snap.isClosed() {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	if ss.inUse {
+		return nil, raft.ErrSnapshotTemporarilyUnavailable
+	}
+
+	if ss.snap == nil {
+		// create snapshot asynchronously
+		ss.createOnce.Do(func() { go ss.createSnap() })
 		return nil, raft.ErrSnapshotTemporarilyUnavailable
 	}
 
+	ss.inUse = true
+	// give transporter the generated snapshot that is ready to send out
+	ss.tr.SnapshotReady(ss.snap, ss.snap.raft().Metadata.Index)
+	return ss.snap, nil
+}
+
+// reportSnap reports the stored snapshot is used up, and clears the snapshot
+// from the snapshot store.
+func (ss *snapshotStore) reportSnap() {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	ss.clearSnap()
+}
+
+// closeSnapBefore closes the stored snapshot if its index is not greater
+// than the given compact index.
+// If it closes the snapshot, it returns true.
+func (ss *snapshotStore) closeSnapBefore(compacti uint64) bool {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+	if ss.snap != nil && ss.snap.raft().Metadata.Index <= compacti {
+		ss.snap.Close()
+		return true
+	}
+	return false
+}
+
+// createSnap creates a new snapshot and stores it into the snapshot store.
+// It also sets a timer to clear the snapshot if it is not in use after
+// some time interval.
+// It should only be called in snapshotStore functions.
+func (ss *snapshotStore) createSnap() {
 	// ask to generate v2 snapshot
 	ss.reqsnapc <- struct{}{}
 	// generate KV snapshot
 	kvsnap := ss.kv.Snapshot()
 	raftsnap := <-ss.raftsnapc
-	ss.snap = newSnapshot(raftsnap, kvsnap)
-	// give transporter the generated snapshot that is ready to send out
-	ss.tr.SnapshotReady(ss.snap, raftsnap.Metadata.Index)
-	return ss.snap, nil
+	snap := newSnapshot(raftsnap, kvsnap)
+
+	ss.mu.Lock()
+	ss.snap = snap
+	ss.mu.Unlock()
+
+	time.AfterFunc(closeUnusedSnapshotInterval, func() {
+		ss.mu.Lock()
+		defer ss.mu.Unlock()
+		if snap == ss.snap && !ss.inUse {
+			ss.clearSnap()
+		}
+	})
+}
+
+// clearSnap clears the snapshot from the snapshot store.
+// It should only be called in snapshotStore functions.
+func (ss *snapshotStore) clearSnap() {
```

Comment:
No. clearSnap clears the snapshot from the snapshot store. It resets related variables and is ready for creating new snapshot. On the other side, caller could close the snapshot to cancel the snapshot, so the one that is using it will error and finish using snapshot. It is confusing because three concepts cross together, and I will clarify this more.